rgw/sts: enhancements and improvements to JWT validation

pritha-srivastava opened the pull request:
in AssumeRoleWithWebIdentity.

added code for token validation using x5c, x5u, x5t, x5t#S256, kid
in header and x5u and modulus, exponent (for RSA group of algos)
in JWKS (JSON Web Key Set).

Fixes: https://tracker.ceph.com/issues/51018
Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

mattbenjamin commented on src/rgw/jwt-cpp/jwt.h:
```diff
@@ -233,6 +237,29 @@ namespace jwt {
+				string n_str = base64_decode_url(modulus);
+				string e_str = base64_decode_url(exponent);
+				unsigned char u_n[n_str.length()], u_e[e_str.length()];
+				std::copy(n_str.begin(), n_str.end(), u_n);
```
it's not very important right now, but this copy could probably be avoided?

pritha-srivastava replied:
BN_bin2bn gave a compilation error with char*

mattbenjamin reviewed (approved):
does this leave any further known gaps wrt, e.g., onelogin?

pritha-srivastava commented on src/rgw/jwt-cpp/jwt.h:
```diff
@@ -659,6 +706,9 @@ namespace jwt {
 				: rsa(public_key, private_key, public_key_password, private_key_password, EVP_sha512, "RS512")
 			{}
+
+			rs512(int number, const std::string& modulus, const std::string& exponent) : rsa(EVP_sha512, modulus, exponent, "RS512")
```
I had to add an integer to this constructor to avoid the ambiguous constructor problem, is there a better suggestion for this. This I dont think is elegant.

tchaikov replied:
a tag struct with a good name, maybe? like,
```c++
struct i_am_a_good_name_t {};
rs512(const i_am_a_good_name_t&, int number, const std::string& modulus, const std::string& exponent) 
 : // ...
{}

pritha-srivastava commented:
> does this leave any further known gaps wrt, e.g., onelogin?

@mattbenjamin : I have tried to cover almost everything here from what I could understand of the JWT spec. The ECDSA keys could also come in like the way we have for RSA (like n,e). But I could not find enough information to compute the public key using those params. Since RSA is a more commonly used algorithm, I have addressed that here. I will open a separate PR for the ECDSA algos when I have adequate information. But we should still encourage the IDPs to upload a cert either as x5c or x5u, because in CreateOpenIDConnectProvider, we save the thumbprint of the cert, which we use to verify the certs in x5c and x5u. If we use params such as n, e, we can not use the thumbprint to validate the public key.

mdw-at-linuxbox commented:
I'll put a few specifics in - but 2 general comments:
1/ the cert validation doesn't look right - looks like it's just validating the consistency of the thumbprints
2/ you might consider using rapidjson.

mdw-at-linuxbox commented on src/rgw/jwt-cpp/jwt.h:
```diff
@@ -11,6 +11,10 @@
+#include <openssl/bn.h>
+#include <openssl/rsa.h>
+
+#include "../rgw_b64.h"
```
style nit: I don't like ".." in include filenames.  Might be best to call this "rgw-jwt.h" instead of "jwt-cpp/jwt.h".

mattbenjamin replied:
+1

mdw-at-linuxbox commented on src/rgw/rgw_rest_sts.cc:
```diff
@@ -188,20 +189,20 @@ WebTokenEngine::get_from_jwt(const DoutPrefixProvider* dpp, const std::string& t
       throw -EACCES;
     }
-    ldpp_dout(dpp, 5) << "Invalid JWT token" << dendl;
+    ldpp_dout(dpp, 5) << "Invalid JWT" << dendl;
```
This debug statement looks exactly like the one on line 196.  Is the error code logged elsewhere?  If not, might be helpful to write it out here.

mdw-at-linuxbox commented on src/rgw/rgw_rest_sts.cc:
```diff
@@ -188,20 +189,20 @@ WebTokenEngine::get_from_jwt(const DoutPrefixProvider* dpp, const std::string& t
   }
   catch (...) {
-    ldpp_dout(dpp, 5) << "Invalid JWT token" << dendl;
+    ldpp_dout(dpp, 5) << "Invalid JWT" << dendl;
```
This debug statement looks exactly like the one on line 192.  It would almost certainly be helpful to capture the exception and stringify it out in the message.

pritha-srivastava commented:
> I'll put a few specifics in - but 2 general comments:
> 1/ the cert validation doesn't look right - looks like it's just validating the consistency of the thumbprints
> 2/ you might consider using rapidjson.
I'll explain 1. The cert validation using thumbprints is based on the fact that CreateOpenIdconnectProvider call in RGW takes a list of thumbprints that the IDP will be using to sign the JWT. And later in RGW when the JWT is received the thumbprint of x5c is calculated and verified against the registered thumbprints. If a match is found, then the incoming x5c cert is used to calculate the signature of the token according to the 'algo' in the JWT and the calculated signature is verified against the incoming signature. Everything that I have read online tells me that this is how we can validate the thumbprints - if you know of anything else that we can do with the thumbprint then please let me know.

mdw-at-linuxbox commented:
My comment re x5c is based on RFC # 7515 which tells me to look at RFC # 5280 for how to validate a certificate chain.  There's nothing in there about "thumbprints".  What is in there is all about validating a chain of signatures ending in a trust anchor, + some mumbly-foo about making sure issuer/subject match up.  { this is the same algorithm used routinely in https to validate certificate chains. }

There is stuff in the openid world about using thumbprints, which at first blush seems somewhat hand-wavey.  Some of it, including the aws doc I found, appears to be describing how to register trust anchors.  Some of it appears to be using thumbprints in a more generic sense.  I gather the rest interface you provide, CreateOpenIDConnectProvider, is intended to register that trust anchor.  (the doc you have for that seems a tad sketchy.)  What's not obvious to me is if you are correctly handling the case where you have one or more intermediate certificates between your trust anchor and the jwt.

Kallio commented on src/rgw/rgw_rest_sts.cc:
```diff
@@ -223,137 +224,265 @@ WebTokenEngine::get_cert_url(const string& iss, const DoutPrefixProvider *dpp, o
+      cert = it;
+    }
+    if (is_cert_valid(thumbprints, cert)) {
+      found_valid_cert = true;
```
found_valid_cert = get true value if we enter the previous loop?  and in line 278 again it will be true?

adamemerson replied:
Can you elaborate on that? I don't think I see the issue. It looks like it's false when we enter the loop and breaks out as soon as it's set true, where it will be at 278.

mdw-at-linuxbox replied:
In the version he was looking at, there was an extra line "found_valid_cert = true;" at the end of the loop.  That means if there was more than one cert (likely always the case) it would always think it had found a trust anchor.  The latest version that pritha pushed removed that line.

If I understand the code that's there now, it will correctly validate that one of the certificates is a trust anchor, but it doesn't validate the chain to see if each signed the previous certificate.    If that's what it's doing, then this logic could be fooled by supplying cert[0] = bad-guy, cert[1]=trusted, cert[0] not signed by cert[1].

adamemerson commented on src/rgw/rgw_rest_sts.cc:
```diff
@@ -223,137 +224,264 @@ WebTokenEngine::get_cert_url(const string& iss, const DoutPrefixProvider *dpp, o
+
+      verifier.verify(decoded);
+    }
+  } catch (std::runtime_error& e) {
```
Why std::runtime_error here instead of std::exception?

Also catch by const reference `(const std::runtime_error& e)`. A bit of a nitpick. At one time it was thought to increase exception performance but on modern compilers it doesn't seem to make a difference.

adamemerson commented on src/rgw/rgw_rest_sts.cc:
```diff
@@ -223,137 +224,264 @@ WebTokenEngine::get_cert_url(const string& iss, const DoutPrefixProvider *dpp, o
 
+    string jwks_url = get_jwks_url(iss, dpp, y);
+    if (jwks_url.empty()) {
+      throw -EINVAL;
```
At some point I want to find all the places that are expecting to catch integers and make them stop. (Though this isn't really your problem since it was like that when you got it.)

adamemerson reviewed (changes requested):
The issue Marcus raised in response to my question.

mattbenjamin commented:
> 
> 
> My comment re x5c is based on RFC # 7515 which tells me to look at RFC # 5280 for how to validate a certificate chain. There's nothing in there about "thumbprints". What is in there is all about validating a chain of signatures ending in a trust anchor, + some mumbly-foo about making sure issuer/subject match up. { this is the same algorithm used routinely in https to validate certificate chains. }
> 
> There is stuff in the openid world about using thumbprints, which at first blush seems somewhat hand-wavey. Some of it, including the aws doc I found, appears to be describing how to register trust anchors. Some of it appears to be using thumbprints in a more generic sense. I gather the rest interface you provide, CreateOpenIDConnectProvider, is intended to register that trust anchor. (the doc you have for that seems a tad sketchy.) What's not obvious to me is if you are correctly handling the case where you have one or more intermediate certificates between your trust anchor and the jwt.

@marcus:  could we potentially well-define the scenario where there is no chain of intermediates, and restrict validation to that case?

scolphoy commented:
I tried this for authenticating against an openid-server that does not expose an `x5c` field, but does expose `n` and `e`, and have so far been unsuccessful. I'm using the boto3 client with python3, and `assume_role_with_web_identity` on client side returns with `botocore.exceptions.ClientError: An error occurred (Unknown) when calling the AssumeRoleWithWebIdentity operation: Unknown`. From the Radosgw log it seems that it starts processing the request the right way, finding the issuer from the id token, finding the correct jwks-endpoint from the issuer's `.well-known/openid-configuration`, finding the `n` and `e` in the keysets, but then suddenly I see parts of some of the trusted CA certificates, some broken characters, and then it apparently check aborts with `Failed the auth strategy, reason=-13`. I have not yet managed to isolate where my problem is.

pritha-srivastava commented:
> I tried this for authenticating against an openid-server that does not expose an `x5c` field, but does expose `n` and `e`, and have so far been unsuccessful. I'm using the boto3 client with python3, and `assume_role_with_web_identity` on client side returns with `botocore.exceptions.ClientError: An error occurred (Unknown) when calling the AssumeRoleWithWebIdentity operation: Unknown`. From the Radosgw log it seems that it starts processing the request the right way, finding the issuer from the id token, finding the correct jwks-endpoint from the issuer's `.well-known/openid-configuration`, finding the `n` and `e` in the keysets, but then suddenly I see parts of some of the trusted CA certificates, some broken characters, and then it apparently check aborts with `Failed the auth strategy, reason=-13`. I have not yet managed to isolate where my problem is.

This is strange, because the n and e computation has been tested by me and and elsewhere and have succeeded in both the places. Please paste snippets of log here if it is possible for you. I am not sure what it means to see 'parts of untrusted CA certificates'.

pritha-srivastava commented:
The 2nd commit will be squashed once the changes are approved.

mdw-at-linuxbox commented:
The only case where x5c.size() == 1 would be valid is if it had a self-signed certificate.  People commonly do that for popup dev environments, but I would not expect to see that in an enterprise environment.  The only other way to avoid validating the certificate chain would be to require that x5c[0] be the trust anchor.  That could work for small to medium scale deployments, but it's going to require due diligence to replace those footprints when they change, which is almost certainly at least yearly (when the certificate expires) .  A large scale deployment of openidconnect (google?) seems very likely to have x5c.size() > 1 with a more or less transient x5c[0], so that's the use case you'd likely exclude by punting on certificate chain validation.

The reason I mentioned intermediate certificates is because I don't fully understand what this code is actually doing.  It may be validating x5c[0] against the system's capath (or some such).  That is certainly not obvious - but since the documentation for registering thumbprints doesn't actually claim to be registering a trust anchor I suppose one could argue it's a documentation problem.  But, since the logic doesn't pass x5c[1+] to the certificate validation process, if there's an intermediate certificate involved between x5c[0] & the system capath, this looks like it has to fail.   If the code is structured right, the same loop that handles intermediate certificates should handle all the other cases (with no more complexity than just handling the smplier cases), so it was the simplest way I could find to exemplify how "this code looks wrong".

pritha-srivastava commented:
> The only case where x5c.size() == 1 would be valid is if it had a self-signed certificate. People commonly do that for popup dev environments, but I would not expect to see that in an enterprise environment. The only other way to avoid validating the certificate chain would be to require that x5c[0] be the trust anchor. That could work for small to medium scale deployments, but it's going to require due diligence to replace those footprints when they change, which is almost certainly at least yearly (when the certificate expires) . A large scale deployment of openidconnect (google?) seems very likely to have x5c.size() > 1 with a more or less transient x5c[0], so that's the use case you'd likely exclude by punting on certificate chain validation.
> 
> The reason I mentioned intermediate certificates is because I don't fully understand what this code is actually doing. It may be validating x5c[0] against the system's capath (or some such). That is certainly not obvious - but since the documentation for registering thumbprints doesn't actually claim to be registering a trust anchor I suppose one could argue it's a documentation problem. But, since the logic doesn't pass x5c[1+] to the certificate validation process, if there's an intermediate certificate involved between x5c[0] & the system capath, this looks like it has to fail. If the code is structured right, the same loop that handles intermediate certificates should handle all the other cases (with no more complexity than just handling the smplier cases), so it was the simplest way I could find to exemplify how "this code looks wrong".

This code is not completely wrong, but yes it does not validate a chain of certs, it only validates a cert when x5c[0] is the trust anchor. As discussed with @mattbenjamin , can I target that bit of code for another PR after opening a tracker issue for the same, and have this merged after running tests, unless @scolphoy has further comments on the modulus and exponent part. This PR was mainly targeted at token validation using modulus and exponent. And I have addressed other comments related to styling and log messages.

scolphoy commented:
Hi, sorry for taking a while to get back to this. Here's some log of a failed assume_role_with_web_identity. I cleaned my id token from the log, but originally it showed as parsed properly. Towards the end of the log, some certificate begin- and end-lines appear abruptly, they seem to match with those found in `/etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt`, which I believe openssl is using while trying to verify something in our oidc-servers response.

Here's the log:
[scolphoy-openidassumerolewithwebid-error.log](https://github.com/ceph/ceph/files/6662658/scolphoy-openidassumerolewithwebid-error.log)

pritha-srivastava commented:
> Hi, sorry for taking a while to get back to this. Here's some log of a failed assume_role_with_web_identity. I cleaned my id token from the log, but originally it showed as parsed properly. Towards the end of the log, some certificate begin- and end-lines appear abruptly, they seem to match with those found in `/etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt`, which I believe openssl is using while trying to verify something in our oidc-servers response.
> 
> Here's the log:
> [scolphoy-openidassumerolewithwebid-error.log](https://github.com/ceph/ceph/files/6662658/scolphoy-openidassumerolewithwebid-error.log)

@scolphoy : I checked the logs and I can see this line: Cert URL is: https://test-user-auth.csc.fi/idp/profile/oidc/keyset. If you check the code here, I have removed the line 'Cert URL is': and replaced it with 'JWKS URL is:'. Are you sure that you used this PR for testing e, and n? I think you still have the old code.

scolphoy commented:
Thanks for checking the log. I was pretty sure, but I guess I've messed up somewhere. I'll start over and try to get back tomorrow.

scolphoy commented:
I tried again, pulled the latest master repo, applied this pull request, recompiled and now it seems that my assume_role_with_web_identity is working; seems I did mess up something in the earlier build. Thanks!

pritha-srivastava commented:
> web_identity is working; seems I did mess up something in the earlier build. Thanks!

Thanks for verifying!

scolphoy commented:
At the end of last week I noticed that in a setup that uses n and e instead of the x5c field, expired tokens were being accepted. I started looking into it and found that just about any token that references a previously added id-provider/issuer would get accepted. I could alter the payload in almost any way I like, and even cut off the signature part of the token (leaving the dot) and it still worked for assuming a role.

I took a look at the source code to see if I can figure out what goes wrong. The validation function throws an exception if the validation fails, and this exception is caught and handled. In the x5c-version of the handler there is a throw -clause, which is missing in the n,e-version of the handler. It seems that the code in the n,e -version might know that the validation failed but the escalation clause is missing, so the execution continues after the catch-block and the token gets accepted by default.

I added the throw clause and am now recompiling to see if that fixes the issue.

pritha-srivastava commented:
> At the end of last week I noticed that in a setup that uses n and e instead of the x5c field, expired tokens were being accepted. I started looking into it and found that just about any token that references a previously added id-provider/issuer would get accepted. I could alter the payload in almost any way I like, and even cut off the signature part of the token (leaving the dot) and it still worked for assuming a role.
> 
> I took a look at the source code to see if I can figure out what goes wrong. The validation function throws an exception if the validation fails, and this exception is caught and handled. In the x5c-version of the handler there is a throw -clause, which is missing in the n,e-version of the handler. It seems that the code in the n,e -version might know that the validation failed but the escalation clause is missing, so the execution continues after the catch-block and the token gets accepted by default.
> 
> I added the throw clause and am now recompiling to see if that fixes the issue.

I can see the missing throw, but I wonder why it wasnt caught by the catch with std:exception. I will check this.

scolphoy commented:
I just found that in `validate_signature_using_n_e` the order of parameters doesn't match between the function definition and how it gets called. Fixing the order now makes my known bad tokens fail validation and refuse giving me the role. I also finally get log output that tells me we're trying to validate RSA signatures and also why the validation fails.

However, my supposedly valid tokens now fail too, with this error:
```
2021-06-21T20:28:27.219+0000 7f796dd1d700  0 req 6 1.072021842s sts:assume_role_web_identity x5c not present
2021-06-21T20:28:27.219+0000 7f796dd1d700  0 req 6 1.072021842s sts:assume_role_web_identity x5u not present
2021-06-21T20:28:27.219+0000 7f796dd1d700  0 req 6 1.072021842s sts:assume_role_web_identity Signature validation using n, e failed: evp verify final failed: 0 error:04091077:rsa routines:int_rsa_verify:wrong signature length
2021-06-21T20:28:27.219+0000 7f796dd1d700 20 req 6 1.072021842s sts:assume_role_web_identity rgw::auth::sts::WebTokenEngine denied with reason=-13
2021-06-21T20:28:27.219+0000 7f796dd1d700  5 req 6 1.072021842s sts:assume_role_web_identity Failed the auth strategy, reason=-13
```

I'm not sure yet if the problem is still with the code or with the signatures in my tokens, but will try to figure out more.

pritha-srivastava commented:
> I just found that in `validate_signature_using_n_e` the order of parameters doesn't match between the function definition and how it gets called. Fixing the order now makes my known bad tokens fail validation and refuse giving me the role. I also finally get log output that tells me we're trying to validate RSA signatures and also why the validation fails.
> 
> However, my supposedly valid tokens now fail too, with this error:
> 
> ```
> 2021-06-21T20:28:27.219+0000 7f796dd1d700  0 req 6 1.072021842s sts:assume_role_web_identity x5c not present
> 2021-06-21T20:28:27.219+0000 7f796dd1d700  0 req 6 1.072021842s sts:assume_role_web_identity x5u not present
> 2021-06-21T20:28:27.219+0000 7f796dd1d700  0 req 6 1.072021842s sts:assume_role_web_identity Signature validation using n, e failed: evp verify final failed: 0 error:04091077:rsa routines:int_rsa_verify:wrong signature length
> 2021-06-21T20:28:27.219+0000 7f796dd1d700 20 req 6 1.072021842s sts:assume_role_web_identity rgw::auth::sts::WebTokenEngine denied with reason=-13
> 2021-06-21T20:28:27.219+0000 7f796dd1d700  5 req 6 1.072021842s sts:assume_role_web_identity Failed the auth strategy, reason=-13
> ```
> 
> I'm not sure yet if the problem is still with the code or with the signatures in my tokens, but will try to figure out more.

@scolphoy : Thank you for identifying the issue! I have found why the signature error you are seeing even after correcting the order of the params. In method setModulusExponentCalcPublicKey, the modulus has not been assigned correctly. After doing that the code works fine for expired and non expired tokens. You can test these changes now or wait till my EOD, I will push another fix here with all the changes ( I have introduced a lot of bugs while restructuring code), so I want to make sure that the code works fine for all scenarios.

scolphoy commented:
Awesome; I tested with fixing that setModulusExponentCalcPublicKey and recompiled; now my known bad and my expired tokens are getting refused and the fresh valid tokens are getting accepted correctly. Good job and thank you @pritha-srivastava

pritha-srivastava commented:
> Awesome; I tested with fixing that setModulusExponentCalcPublicKey and recompiled; now my known bad and my expired tokens are getting refused and the fresh valid tokens are getting accepted correctly. Good job and thank you @pritha-srivastava

@scolphoy : Can you please review the last commit that I have added, since you know of all the issues and the fixes.

scolphoy commented:
> @scolphoy : Can you please review the last commit that I have added, since you know of all the issues and the fixes. I did test them by running the code through gdb to make sure the signature validation code is actually invoked!

Sure. Looks good to me. Also compiled and tested with my tokens and it works the way I expected. :)

scolphoy commented:
I just figured out one more thing that we should add. The non-x5c -version currently does not do anything with JWK thumbprints. When adding an OpenID provider, we are adding a thumbprint, but that gets ignored in the non-x5c -version; I verified this by setting my providers thumbprint to just 40 A-letters and my valid tokens it still worked.

Currently we are using a list of Arns to tell whick OpenIDConnect providers we should trust. When we receive a token with an issuer url that matches one of the providers on that list, we fetch the jwks_uri, find the n and e, construct the public key, verify the signature with that public key and we can be sure that the token was signed by the provider that answers from that url. This is very good. However, currently we can't tell if the provider that answers from that url is the one that we wanted.

One scenario of when this could become a problem is that we add an openid-provider, set and forget, eventually the domain in the url expires and someone else buys it. They can now set up a new openid server with their own new signing keys, and because we are checking only the url and that the signature matches the keys from the url, we will now be trusting the tokens that they issue. Changing keys would also change the thumbprint, so simply controlling the domain in the url would no longer be enough to compromise trust.

pritha-srivastava commented:
> I just figured out one more thing that we should add. The non-x5c -version currently does not do anything with JWK thumbprints. When adding an OpenID provider, we are adding a thumbprint, but that gets ignored in the non-x5c -version; I verified this by setting my providers thumbprint to just 40 A-letters and my valid tokens it still worked.
> 
> Currently we are using a list of Arns to tell whick OpenIDConnect providers we should trust. When we receive a token with an issuer url that matches one of the providers on that list, we fetch the jwks_uri, find the n and e, construct the public key, verify the signature with that public key and we can be sure that the token was signed by the provider that answers from that url. This is very good. However, currently we can't tell if the provider that answers from that url is the one that we wanted.
> 
> One scenario of when this could become a problem is that we add an openid-provider, set and forget, eventually the domain in the url expires and someone else buys it. They can now set up a new openid server with their own new signing keys, and because we are checking only the url and that the signature matches the keys from the url, we will now be trusting the tokens that they issue. Changing keys would also change the thumbprint, so if we also verify the thumbprints, simply controlling the domain in the url would no longer be enough to compromise trust.

Yes I am aware that we do not use the thumbprint of the x5c certificate. Do you know how to verify the x5c cert thumbprint with the public key? Or are you suggesting that we store the public key thumbprint in the CreateOpenIdConnectProvider call? AWS says that we need to save thumbprint of X.509 certificate (https://docs.aws.amazon.com/IAM/latest/APIReference/API_CreateOpenIDConnectProvider.html)

scolphoy commented:
I was suggesting storing the thumbprint in the CreateOpenIdConnectProvider call, yes.

And you're right, looks like AWS is not using JWK thumbprints either. It seems that instead they are using thumbprints of the X509 certificate that is used to authenticate the HTTPS connection to the ID provider's endpoint, i.e. they seem to be pinning what is a trusted channel for delivering the public keys, rather than pinning the public keys themselves. I suppose this works too, just that these certificates tend to get renewed relatively frequently these days and would have to be updated to rgw each time.

What I was actually thinking about was using JWK thumbprints, which can be derived from the e, kty and n fields of a key, storing the thumbprint when the id provider is added, and calculating and comparing the thumbprint again when fetching the jwks_keys as part of verifying a token. I'm not sure if it would be possible to derive this from the x5c field. These would pin us to the keys used for signing the tokens, rather than to the connection through which those keys are delivered. But since this doesn't seem to be what AWS is doing after all, if we want to be compatible, perhaps we shouldn't do it now either.

There is no guarantee that the X509 certificate used to authenticate the connection and the X509 certificate used to sign the tokens would be the same certificate, especially since you can serve tokens for many realms (with hopefully separate keys) from the same id provider. I believe computing the thumbprint from the x5c field would be wrong. Instead the correct X509 certificate should be found from the parameters of the underlying HTTPS connection when connecting to the the jwks_keys endpoint and the thumbprint computed from that. With a quick search, I can find shell scripts that do this (e.g. [here](https://gist.github.com/riccardomc/a71e14bf9c9a45632185a1445ef1ee03)), but how to integrate it properly in rgw would need more work.

pritha-srivastava commented:
> I was suggesting storing the thumbprint in the CreateOpenIdConnectProvider call, yes.
> 
> And you're right, looks like AWS is not using JWK thumbprints either. It seems that instead they are using thumbprints of the X509 certificate that is used to authenticate the HTTPS connection to the ID provider's endpoint, i.e. they seem to be pinning what is a trusted channel for delivering the public keys, rather than pinning the public keys themselves. I suppose this works too, just that these certificates tend to get renewed relatively frequently these days and would have to be updated to rgw each time.
> 
> What I was actually thinking about was using JWK thumbprints, which can be derived from the e, kty and n fields of a key, storing the thumbprint when the id provider is added, and calculating and comparing the thumbprint again when fetching the jwks_keys as part of verifying a token. I'm not sure if it would be possible to derive this from the x5c field. These would pin us to the keys used for signing the tokens, rather than to the connection through which those keys are delivered. But since this doesn't seem to be what AWS is doing after all, if we want to be compatible, perhaps we shouldn't do it now either.
> 
I had checked and couldn't find deriving the thumbprint from x5c for a public key to be possible (if i am understanding correctly what you mean), I will double check again. Meanwhile I will also discuss with my team members if we can store the thumbprint of the public key while creating a openidconnectprovider in RGW (after all its a thumbprint).

> There is no guarantee that the X509 certificate used to authenticate the connection and the X509 certificate used to sign the tokens would be the same certificate, especially since you can serve tokens for many realms (with hopefully separate keys) from the same id provider. I believe computing the thumbprint from the x5c field would be wrong. Instead the correct X509 certificate should be found from the parameters of the underlying HTTPS connection when connecting to the the jwks_keys endpoint and the thumbprint computed from that. With a quick search, I can find shell scripts that do this (e.g. [here](https://gist.github.com/riccardomc/a71e14bf9c9a45632185a1445ef1ee03)), but how to integrate it properly in rgw would need more work.

If you are talking about the current x5c cert validation - yes there is a gap. Right now it only addresses the case when there is a single cert which is the cert that signs the tokens and the root CA. I plan to open a tracker issue and address this properly in another PR which takes the thumbprint and validates the certificate chain in the incoming token according to the RFC. Other reviewers have raised this point here.

scolphoy commented:
I checked the other reviewers comments, and what I'm talking about is something else. I'm not talking about validating the certificate chain in the x5c field.

As I see it, there are three kinds of thumbprints involved.
1. The X509 thumbprint one can compute from a certificate in the x5c chain
2. The JWK thumbprint one can compute from e, kty and n
3. The X509 thumbprint one can compute from the X509 certificate (the SSL/TLS-certificate) used for the https connection to the jwks_keys endpoint that delivers all the other keys and certificates

Types 1 and 2 are used to pin to the keys used to verify tokens, the type 3 is used to pin to the delivery channel for keys used to verify tokens, but not the keys themselves. Type 1 is possible only for providers that provide the x5c field. Type 2 is available with or without the x5c field; if x5c is provided it is expected to be configured properly, so that the key in the root certificate matches the key made of e and n. JWK thumbprints are of different length than X509 thumbprints, so our current "exactly 40 characters" in the thumbprints would not support JWK thumbprints.

Type 3 is the lowest common denominator that works regardless of what fields our jwks_keys endpoint supports, if the x5c field is there or not, what signature scheme (if any) we use, and so on. It pins the delivery of the verification keys saying "trust whatever this https endpoint says, but only if it's secured using these certificates in particular" and means that if someone later gains rightful access to the domain and gets a new valid certificate for it, the thumbprint-check would be the last line of defence that prevents us from accidentally trusting them.

Judging by [AWS instructions](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc_verify-thumbprint.html), seeing how they are using an X509 thumbprint and how they are not extracting the x5c field in any of the steps, but are computing a fingerprint from one of the certificates returned then connecting to the jwks_keys https endpoint, they must be using type 3. If we want to do it the same way as AWS, this is what we should do too. The particular certificates used in type 1 and type 3 are very likely different, and that's why using the x5c field for computing the thumbprint would very likely cause at least the above AWS instructions to fail in many cases.

pritha-srivastava commented:
> I checked the other reviewers comments, and what I'm talking about is something else. I'm not talking about validating the certificate chain in the x5c field.
> 
> As I see it, there are three kinds of thumbprints involved.
> 
> 1. The X509 thumbprint one can compute from a certificate in the x5c chain
> 2. The JWK thumbprint one can compute from e, kty and n
> 3. The X509 thumbprint one can compute from the X509 certificate (the SSL/TLS-certificate) used for the https connection to the jwks_keys endpoint that delivers all the other keys and certificates
> 
> Types 1 and 2 are used to pin to the keys used to verify tokens, the type 3 is used to pin to the delivery channel for keys used to verify tokens, but not the keys themselves. Type 1 is possible only for providers that provide the x5c field. Type 2 is available with or without the x5c field; if x5c is provided it is expected to be configured properly, so that the key in the root certificate matches the key made of e and n. JWK thumbprints are of different length than X509 thumbprints, so our current "exactly 40 characters" in the thumbprints would not support JWK thumbprints.
> 
> Type 3 is the lowest common denominator that works regardless of what fields our jwks_keys endpoint supports, if the x5c field is there or not, what signature scheme (if any) we use, and so on. It pins the delivery of the verification keys saying "trust whatever this https endpoint says, but only if it's secured using these certificates in particular" and means that if someone later gains rightful access to the domain and gets a new valid certificate for it, the thumbprint-check would be the last line of defence that prevents us from accidentally trusting them.
> 
> Judging by [AWS instructions](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc_verify-thumbprint.html), seeing how they are using an X509 thumbprint and how they are not extracting the x5c field in any of the steps, but are computing a fingerprint from one of the certificates returned when connecting to the jwks_keys https endpoint, they must be using type 3. If we want to do it the same way as AWS, this is what we should do too. The particular certificates used in type 1 and type 3 are very likely different, and that's why using the x5c field for computing the thumbprint would very likely cause at least the above AWS instructions to fail in many cases.

So if there is a single certificate in the certificate chain (like self signed certs) 1 and 3 become the same and that is what has been addressed here in RGW so far. ( I have checked using the instructions given in AWS documentation). As I told before we know this is a gap, and we intend to address this in another PR where we will cover other cases when we store the root CA thumbprint (following AWS instructions) while creating an openidconnectprovider in rgw, and then later when we verify the x5c certs in the incoming token, we need to check the thumbprint of the root CA in the x5c certs as well as verify each intermediate certificate in the certificate chain.

Meanwhile, I will add the thumbprint verification for n and e using n, kty and e to this PR.

pritha-srivastava commented:
> > I checked the other reviewers comments, and what I'm talking about is something else. I'm not talking about validating the certificate chain in the x5c field.
> > As I see it, there are three kinds of thumbprints involved.
> > 
> > 1. The X509 thumbprint one can compute from a certificate in the x5c chain
> > 2. The JWK thumbprint one can compute from e, kty and n
> > 3. The X509 thumbprint one can compute from the X509 certificate (the SSL/TLS-certificate) used for the https connection to the jwks_keys endpoint that delivers all the other keys and certificates
> > 
> > Types 1 and 2 are used to pin to the keys used to verify tokens, the type 3 is used to pin to the delivery channel for keys used to verify tokens, but not the keys themselves. Type 1 is possible only for providers that provide the x5c field. Type 2 is available with or without the x5c field; if x5c is provided it is expected to be configured properly, so that the key in the root certificate matches the key made of e and n. JWK thumbprints are of different length than X509 thumbprints, so our current "exactly 40 characters" in the thumbprints would not support JWK thumbprints.
> > Type 3 is the lowest common denominator that works regardless of what fields our jwks_keys endpoint supports, if the x5c field is there or not, what signature scheme (if any) we use, and so on. It pins the delivery of the verification keys saying "trust whatever this https endpoint says, but only if it's secured using these certificates in particular" and means that if someone later gains rightful access to the domain and gets a new valid certificate for it, the thumbprint-check would be the last line of defence that prevents us from accidentally trusting them.
> > Judging by [AWS instructions](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc_verify-thumbprint.html), seeing how they are using an X509 thumbprint and how they are not extracting the x5c field in any of the steps, but are computing a fingerprint from one of the certificates returned when connecting to the jwks_keys https endpoint, they must be using type 3. If we want to do it the same way as AWS, this is what we should do too. The particular certificates used in type 1 and type 3 are very likely different, and that's why using the x5c field for computing the thumbprint would very likely cause at least the above AWS instructions to fail in many cases.
> 
> So if there is a single certificate in the certificate chain (like self signed certs) 1 and 3 become the same and that is what has been addressed here in RGW so far. ( I have checked using the instructions given in AWS documentation). As I told before we know this is a gap, and we intend to address this in another PR where we will cover other cases when we store the root CA thumbprint (following AWS instructions) while creating an openidconnectprovider in rgw, and then later when we verify the x5c certs in the incoming token, we need to check the thumbprint of the root CA in the x5c certs as well as verify each intermediate certificate in the certificate chain.
> 
> Meanwhile, I will add the thumbprint verification for n and e using n, kty and e to this PR.

Or are you saying that even for n,e we can store the thumbprint according to AWS documentation, and then we can use the jwks_uri and compute the thumbprint and validate it against the one stored?

scolphoy commented:
Yes, even with n,e, we can use the AWS documentation and what I described as type 3 above.

Also I don't think what you said about type 1 and type 3 becoming the same is generally true. For example, if CSC and Red hat wanted to use a shared OpenID provider, but separate realms, we might want to have one certificate for the https endpoint, one for signing tokens for the CSC realm and one for signing tokens for the Red hat realm. In this case we would have two type 1 certificates/thumbprints and one type 3, and these would all be different. In a single realm setup one might put the same certificate for the token signing and for the web server, in which case they indeed would become the same, but it sounds like a bad practice and we should not assume that anyone does that.

pritha-srivastava commented:
> Yes, even with n,e, we can use the AWS documentation and what I described as type 3 above.
> 
Ok.
> There would be no need to compute using n, e and kty ; unless we want to also lock on to the particular signature scheme and particular keys, instead of just particular endpoint on where to find the keys. It's a tradeoff. Seems AWS is not doing this.
> 
> Also I don't think what you said about type 1 and type 3 becoming the same is generally true. For example, if CSC and Red hat wanted to use a shared OpenID provider, but separate realms, we might want to have one certificate for the https endpoint, one for signing tokens for the CSC realm and one for signing tokens for the Red hat realm. In this case we would have two type 1 certificates/thumbprints and one type 3, and these would all be different. In a single realm setup one might put the same certificate for the token signing and for the web server, in which case they indeed would become the same, but it sounds like a bad practice and we should not assume that anyone does that.
Ok.

scolphoy commented:
Come to think of it, what we now have in this PR is already a significant improvement over what was there before. I'd be happy to see this merged soon. It does now check that the tokens are valid and signed properly. To replace the accepted signing keys from the outside, one would have to redirect the jwks_keys query to their own endpoint and also have a valid, even if not pinned, certificate for the domain so that libcurl will accept to fetch the keys. My biggest concern with this is that someone is going to deploy this and then neglect following the state of the providers on their list properly.

If the thumbprint validation is going to be a separate pull request, I'm ok with that. Though I do maintain that there is a difference between thumbprinting from the x5c chain and thumbprinting form the certificate from the TLS handshake with the endpoint (what AWS does). The latter is more generic, doesn't make us depend on the x5c field being present, and doesn't lock us to a particular key, but to a particular key delivery endpoint.

pritha-srivastava commented:
> Come to think of it, what we now have in this PR is already a significant improvement over what was there before. I'd be happy to see this merged soon. It does now check that the tokens are valid and signed properly. To replace the accepted signing keys from the outside, one would have to redirect the jwks_keys query to their own endpoint and also have a valid, even if not pinned, certificate for the domain so that libcurl will accept to fetch the keys. My biggest concern with this is that someone is going to deploy this and then neglect following the state of the providers on their list properly.
> 
> If the thumbprint validation is going to be a separate pull request, I'm ok with that. Though I do maintain that there is a difference between thumbprinting from the x5c chain and thumbprinting form the certificate from the TLS handshake with the endpoint (what AWS does). The latter is more generic, doesn't make us depend on the x5c field being present, and doesn't lock us to a particular key, but to a particular key delivery endpoint.

I have created a tracker issue here: https://tracker.ceph.com/issues/51346 for tracking the thumbprint validation code. Please feel free to take it up if you can before I get to it. And thanks for all the input.

cbodley commented:
@pritha-srivastava what's left to do here?

pritha-srivastava commented:
> @pritha-srivastava what's left to do here?

Properly validating the certificate chain is still left to be done.

mattbenjamin commented:
@pritha-srivastava ping?
